code/game/machinery/computer/sm_monitor.dm

```diff
@@ -52,23 +52,23 @@
 		data["SM_ambienttemp"] = air.temperature
 		data["SM_ambientpressure"] = air.return_pressure()
 		//data["SM_EPR"] = round((air.total_moles / air.group_multiplier) / 23.1, 0.01)
-		var/other_moles = air.total_trace_moles()
+		var/list/gasdata = list()
 		var/TM = air.total_moles()
 		if(TM)
-			data["SM_gas_O2"] = round(100*air.oxygen/TM, 0.01)
-			data["SM_gas_CO2"] = round(100*air.carbon_dioxide/TM, 0.01)
-			data["SM_gas_N2"] = round(100*air.nitrogen/TM, 0.01)
-			data["SM_gas_PL"] = round(100*air.toxins/TM, 0.01)
-			if(other_moles)
-				data["SM_gas_OTHER"] = round(100 * other_moles / TM, 0.01)
-			else
-				data["SM_gas_OTHER"] = 0
+			gasdata.Add(list(list("name"= "Oxygen", "amount" = round(100 * air.oxygen / TM, 0.01))))
+			gasdata.Add(list(list("name"= "Carbon Dioxide", "amount" = round(100 * air.carbon_dioxide / TM, 0.01))))
+			gasdata.Add(list(list("name"= "Nitrogen", "amount" = round(100 * air.nitrogen / TM, 0.01))))
+			gasdata.Add(list(list("name"= "Plasma", "amount" = round(100 * air.toxins / TM, 0.01))))
+			gasdata.Add(list(list("name"= "Nitrous Oxide", "amount" = round(100 * air.sleeping_agent / TM, 0.01))))
+			gasdata.Add(list(list("name"= "Agent B", "amount" = round(100 * air.agent_b / TM, 0.01))))
```

Comment:
this is gross because of how we handle it vs TG; they just use a for loop....we can't do that....

Reply:
I think first you could define a (static) list at proc or class level, something along the lines of:
```dm
var/static/list/gasses = list(
	"oxygen" = "Oxygen",
	"carbon_dioxide" = "Carbon Dioxide",
	// ETC...
)
```

Then you could use a define along with direct substitution:
```dm
/// To be used only in ui_data()
#define GAS_ENTRY_DATA(n, k) list(list("name" = "[n]", "amount" = round(100 * air.[##k] / TM, 0.01)))
```

...Finally allowing you to make a clean loop in the data proc:
```dm
for(var/gastype in gasses)
	gasdata += GAS_ENTRY_DATA(gasses[gastype], gastype)
```

Reply:
Yeah, I contemplated this, but I'd like to generalize it, especially so it can be utilized with something like `getGasLabel` and `getGasColor` (both TGUI constants), as it'd make looping through every gas type a bit easier.

Unfortunately, in part, it's trying to fit a var based atmospheric system into working with a list based one, so you're going to wind up with some nasty stuff when converting between the two.